x-pack/plugins/monitoring/public/components/logstash/pipeline_viewer/models/pipeline/if_statement.js

```diff
@@ -20,23 +25,29 @@ export class IfStatement extends Statement {
   static fromPipelineGraphVertex(ifVertex, pipelineStage) {
     const trueStatements = [];
     const elseStatements = [];
-
+    const { trueOutgoingVertices } = ifVertex;
+    const { falseOutgoingVertices } = ifVertex;
     const trueVertex = ifVertex.trueOutgoingVertex;
     const falseVertex = ifVertex.falseOutgoingVertex;
     const next = ifVertex.next;
 
     let currentVertex = trueVertex;
     while (isVertexPipelineStage(currentVertex, pipelineStage) && (currentVertex !== next)) {
+      pull(trueOutgoingVertices, currentVertex);
       trueStatements.push(makeStatement(currentVertex, pipelineStage));
       currentVertex = currentVertex.next;
     }
 
     currentVertex = falseVertex;
     while (currentVertex && isVertexPipelineStage(currentVertex, pipelineStage) && (currentVertex !== next)) {
+      pull(falseOutgoingVertices, currentVertex);
       elseStatements.push(makeStatement(currentVertex, pipelineStage));
       currentVertex = currentVertex.next;
     }
 
+    addVertices(trueStatements, trueOutgoingVertices, pipelineStage, makeStatement);
+    addVertices(elseStatements, falseOutgoingVertices, pipelineStage, makeStatement);
+
```

Comment:
I want to check my understanding of the change in this PR and, in particular, this part of the code.

Currently (prior to this PR), we assumed incorrectly that there could be only one `trueVertex` and at most one `falseVertex`. The code here in this PR fixes that assumption. Specifically we now want to get _all_ true vertices and all false vertices, and add the statements the result in to `trueStatements` and `elseStatements`, respectively. Am I understanding this right?

If so, could this alternative code work instead?

```js
const { trueOutgoingVertices, falseOutgoingVertices } = ifVertex;
const next = ifVertex.next;

trueOutgoingVertices.forEach(trueVertex => {
  const currentVertex = trueVertex;
  while(isVertexPipelineStage(currentVertex, pipelineStage) && (currentVertex !== next)) {
    trueStatements.push(makeStatement(currentVertex, pipelineStage));
    currentVertex = currentVertex.next;
  }
}

falseOutgoingVertices.forEach(falseVertex => {
  const currentVertex = falseVertex;
  while(isVertexPipelineStage(currentVertex, pipelineStage) && (currentVertex !== next)) {
    elseStatements.push(makeStatement(currentVertex, pipelineStage));
    currentVertex = currentVertex.next;
  }
}
```

If it does work, what do you think of it? I found it simpler to reason about, particularly since it doesn't require the side-effecting `pull` operation and since it keeps the adding of statements in one place, but that might be my bias since I came up with it.

Reply:
I agree that I'd like to simplify that code, and I came into this review knowing it'd probably be scrutinized for good reason. I'm 👍 with this solution.
Some thoughts:
- Are we ever concerned about objects not of type `IfVertex` being passed (i.e. should we check `trueOutgoingVertices` and `falseOutgoingVertices` for `undefined` state.)? The existing solution doesn't do that, so I'm assuming it's ok to not perform these checks here, although it means some tests need updated
- These two blocks are logically identical, they're just operating on different collections, we can probably refactor it into a function
- I think we should annotate the `next` prop on the `IfVertex` class (probably in a separate PR); the code accomplishes its task in an elegant way but I feel it's also a little challenging to grok at a glance

Reply:
> Are we ever concerned about objects not of type IfVertex being passed (i.e. should we check trueOutgoingVertices and falseOutgoingVertices for undefined state.)? The existing solution doesn't do that, so I'm assuming it's ok to not perform these checks here, although it means some tests need updated

Yeah, I'd leave that out of this PR for now. I'm not too concerned about this since it's an internal API. We can add these types of checks in a later PR, if we want.

> These two blocks are logically identical, they're just operating on different collections, we can probably refactor it into a function

++.

> I think we should annotate the next prop on the IfVertex class (probably in a separate PR); the code accomplishes its task in an elegant way but I feel it's also a little challenging to grok at a glance

++.